server/src/main/java/org/elasticsearch/cluster/coordination/ClusterBootstrapService.java

```diff
@@ -21,214 +21,188 @@
+        if (unconfiguredBootstrapTimeout == null) {
+            return;
+        }
 
         if (transportService.getLocalNode().isMasterNode() == false) {
             return;
         }
 
-        if (unconfiguredBootstrapTimeout != null) {
-            logger.info("no discovery configuration found, will perform best-effort cluster bootstrapping after [{}] " +
-                    "unless existing master is discovered", unconfiguredBootstrapTimeout);
-            final ThreadContext threadContext = transportService.getThreadPool().getThreadContext();
-            try (ThreadContext.StoredContext ignore = threadContext.stashContext()) {
-                threadContext.markAsSystemContext();
+        logger.info("no discovery configuration found, will perform best-effort cluster bootstrapping after [{}] " +
+            "unless existing master is discovered", unconfiguredBootstrapTimeout);
+
+        transportService.getThreadPool().scheduleUnlessShuttingDown(unconfiguredBootstrapTimeout, Names.GENERIC, new Runnable() {
+            @Override
+            public void run() {
+                final Set<DiscoveryNode> discoveredNodes = getDiscoveredNodes();
+                final List<DiscoveryNode> zen1Nodes = discoveredNodes.stream().filter(Coordinator::isZen1Node).collect(Collectors.toList());
+                if (zen1Nodes.isEmpty()) {
+                    logger.debug("performing best-effort cluster bootstrapping with {}", discoveredNodes);
+                    startBootstrap(discoveredNodes, 0);
+                } else {
+                    logger.info("avoiding best-effort cluster bootstrapping due to discovery of pre-7.0 nodes {}", zen1Nodes);
+                }
+            }
 
-                transportService.getThreadPool().scheduleUnlessShuttingDown(unconfiguredBootstrapTimeout, Names.SAME, new Runnable() {
+            @Override
+            public String toString() {
+                return "unconfigured-discovery delayed bootstrap";
+            }
+        });
+    }
+
+    private Set<DiscoveryNode> getDiscoveredNodes() {
+        return Stream.concat(Stream.of(transportService.getLocalNode()),
+            StreamSupport.stream(discoveredNodesSupplier.get().spliterator(), false)).collect(Collectors.toSet());
+    }
+
+    private void startBootstrap(Set<DiscoveryNode> discoveryNodes, int placeholderCount) {
+        assert discoveryNodes.stream().allMatch(DiscoveryNode::isMasterNode) : discoveryNodes;
+        assert discoveryNodes.stream().noneMatch(Coordinator::isZen1Node) : discoveryNodes;
+        assert placeholderCount < discoveryNodes.size() : discoveryNodes.size() + " <= " + placeholderCount;
+        if (bootstrappingPermitted.compareAndSet(true, false)) {
+            doBootstrap(new VotingConfiguration(Stream.concat(discoveryNodes.stream().map(DiscoveryNode::getId),
+                Stream.generate(() -> BOOTSTRAP_PLACEHOLDER_PREFIX + UUIDs.randomBase64UUID(random)).limit(placeholderCount))
```

Comment:
instead of randomly-generated UUIDs here, I think we should combine the `BOOTSTRAP_PLACEHOLDER_PREFIX` with the `initial_master_nodes` value here. This will make it clearer which entry we have a place-holder for.

Reply:
Very well, I did so in c7167340a30ae878c094f6a63865f257f4d0a94e and also ensured that there are no duplicate requirements in eadc79797854da78bae49381cb8b571ea64270de.